packages/@lwc/template-compiler/src/parser/expression-complex/html.ts

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { parseExpressionAt } from 'acorn';
+import { Document, DocumentFragment, ParsingError, ParserOptions } from 'parse5';
+import { ParserDiagnostics, invariant } from '@lwc/errors';
+import Parser from 'parse5/lib/parser';
+import Tokenizer from 'parse5/lib/tokenizer';
+import type ParserCtx from '../parser';
+import type { PreparsedExpressionMap } from './types';
+
+const OPENING_CURLY_BRACKET = 0x7b;
+const CLOSING_CURLY_BRACKET = 0x7d;
+const WHITESPACE = /(\s*)/;
+
+function getWhitespaceLen(str: string): number {
+    return WHITESPACE.exec(str)![0].length;
+}
+
+/**
+ * This class extends `parse5`'s internal tokenizer.
+ *
+ * Its behavior diverges from that specified in the WHATWG HTML spec
+ * in two places:
+ *   - 13.2.5.38 - unquoted attribute values
+ *   - 13.2.5.1 - the "data" state, which corresponds to parsing outside of tags
+ *
+ * Specifically, this tokenizer defers to Acorn's JavaScript parser when
+ * encountering a `{` character for an attribute value or within a text
+ * node. Acorn parses the expression, and the tokenizer continues its work
+ * following the closing `}`.
+ */
+class TemplateHtmlTokenizer extends Tokenizer {
+    parser: TemplateHtmlParser;
+    // We track which attribute values are in-progess so that we can defer
+    // to the default tokenizer's behavior after the first character of
+    // an unquoted attr value has been checked for an opening curly brace.
+    checkedAttrs = new WeakSet<any>();
+
+    constructor(parser: TemplateHtmlParser) {
+        super();
+        this.parser = parser;
+    }
+
+    parseTemplateExpression() {
+        const expressionStart: number = this.preprocessor.pos;
+        const html = this.preprocessor.html as string;
+
+        const leadingWhitespaceLen = getWhitespaceLen(html.slice(expressionStart + 1));
+        const javascriptExprStart = expressionStart + leadingWhitespaceLen + 1;
+
+        // Start parsing immediately after the opening curly brace.
+        const estreeNode = parseExpressionAt(html, javascriptExprStart, {
+            ecmaVersion: 2022,
```

Comment:
ES2022 will be parsed, but none of the newly introduced syntax is permitted. This will just make for better error messages; the component author will be notified about their use of disallowed syntax rather than an arcane parsing error.

Reply:
Are we going to need to periodically update this? Can we just use [`"latest"`](https://github.com/acornjs/acorn/blob/master/acorn/README.md#interface)? (Which, presumably, will bump when we bump our acorn dep.)

Reply:
Good question. With the current implementation, the AST is validated using a blocklist for disallowed node types. If we used `"latest"` here, we'd probably want to invert that to avoid permitting new JS syntax proposals .

Reply:
Added `TMPL_EXPR_ECMASCRIPT_EDITION` constant to ensure consistency